musicbot/bot.py

```diff
@@ -625,8 +626,12 @@ def _init_player(self, player, *, server=None):
     async def on_player_finished_playing(self, player, **_):
         if not player.playlist.entries and not player.current_entry and self.config.auto_playlist:
             while self.autoplaylist:
-                random.shuffle(self.autoplaylist)
-                song_url = random.choice(self.autoplaylist)
+                random.shuffle(self.song_list)
+                if (len(self.song_list) == 0):
+                    print('All songs have been played. Restarting auto playlist...')
+                    self.song_list = self.autoplaylist[:]
+                song_url = random.choice(self.song_list)
+                self.song_list.remove(song_url)
```

Comment:
As the list is already shuffled (line 629), it may make sense to replace lines 633-634 with:

```py
song_url = self.autoplaylist.pop()
```

See: https://docs.python.org/3.5/tutorial/datastructures.html